### ISSUE TITLE
[Feature Request] Implement Streaming Hint Injection with RL-Trained Retriever

### ISSUE THREAD

**Issue by codelion** [codelion/optillm#206]
## 🎯 Overview
Implement a novel inference optimization approach inspired by [this research idea](https://x.com/willccbb/status/1940557166248972387): **a lightweight retriever that processes streaming Chain-of-Thought reasoning to inject contextual hints from a memory bank, trained using downstream task performance as the reward signal**.

This would make optillm the first framework to support real-time, RL-trained hint injection during LLM reasoning.

## 🔬 Problem Description
Current LLM inference approaches either:
- Provide all context upfront (overwhelming and unfocused)
- Generate responses without external guidance (missing relevant knowledge)
- Use static retrieval (not adaptive to reasoning progress)

**Will Brown's approach** solves this by:
1. **Monitoring** the model's reasoning stream in real-time
2. **Retrieving** relevant hints from a memory bank at strategic moments
3. **Injecting** hints precisely when they're most helpful
4. **Learning** optimal injection strategies through RL using downstream task performance

### Implementation Structure
Create a new folder `optillm/streaming_hints/` (similar to `optillm/autothink/`) containing:

```
optillm/streaming_hints/
├── __init__.py
├── README.md                    # Detailed implementation documentation
├── streaming_hints.py           # Main approach implementation
├── memory_bank.py              # Hint storage and retrieval
├── retriever.py                # Streaming hint injection logic
├── rl_trainer.py               # Reinforcement learning components
└── evaluator.py                # Performance evaluation utilities
```

### Core Components

```python
# optillm/streaming_hints/streaming_hints.py
class StreamingHintsApproach:
    def __init__(self):
        self.SLUG = "streaming_hints"  # or "sh" for short
        self.memory_bank = HintMemoryBank()
        self.retriever = StreamingHintRetriever(self.memory_bank)
        self.trainer = HintRetrievalRLTrainer(self.retriever)
```

### 1. **Hint Memory Bank** (`memory_bank.py`)
- Store curated hints with embeddings
- Support vector similarity search
- Track usage statistics and effectiveness

### 2. **Streaming Hint Retriever** (`retriever.py`)
- Process CoT reasoning token-by-token
- Detect optimal injection points (uncertainty signals, reasoning transitions)
- Retrieve relevant hints using embedding similarity
- Inject hints as `<hint>content</hint>` tags in the stream

### 3. **RL Training System** (`rl_trainer.py`)
- Compare performance with/without hints
- Use downstream task accuracy as reward signal
- Train injection timing and hint selection policies
- Support multiple domains (math, coding, reasoning)

### 4. **Documentation** (`README.md`)
The folder's README should include:
- **Overview**: What streaming hint injection accomplishes
- **Architecture**: How components work together
- **Implementation Details**: Key algorithms and design decisions
- **Usage Examples**: Code samples and configuration options
- **Evaluation Results**: Benchmark performance and comparisons
- **Future Work**: Potential improvements and extensions

## 🎨 Integration with Existing optillm

This builds naturally on existing approaches:
- **Memory Plugin**: Extend for hint storage and retrieval
- **Router Plugin**: Similar classification concept for hint relevance
- **CoT Reflection**: Compatible with structured reasoning sections
- **MCP Plugin**: Shows how to integrate external tools during reasoning

### Usage Examples

```python
# Via model name prefix
model="streaming_hints-gpt-4o-mini"

# Combined with existing approaches
model="streaming_hints&cot_reflection-gpt-4o-mini"  # Pipeline
model="streaming_hints|moa-gpt-4o-mini"             # Parallel

# With custom configuration
extra_body={"optillm_approach": "streaming_hints", "hint_threshold": 0.8}
```

## 📊 Expected Benefits

1. **Performance**: Significant improvements on complex reasoning tasks
2. **Efficiency**: Targeted hint injection vs. overwhelming context
3. **Adaptability**: RL learns optimal strategies for different domains
4. **Composability**: Works with existing optillm approaches
5. **Innovation**: Novel real-time reasoning enhancement

### Benchmark Tasks
- **Math**: GSM8K, MATH dataset problems
- **Coding**: HumanEval, MBPP coding challenges  
- **Reasoning**: LogiQA, ReClor logical reasoning
- **General**: MMLU multi-domain questions

### Evaluation Metrics
- Accuracy improvement vs. baseline
- Token efficiency (performance per token)
- Hint relevance and timing effectiveness

---

**This is a high-impact feature that could significantly advance the field of LLM inference optimization. Looking forward to collaborating with the community to bring this innovative approach to life! 🚀**
